File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    List<List<Integer>> result;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        result = new ArrayList<>();

        if(candidates == null || candidates.length == 0) return result;
        helper(candidates, 0, target, new ArrayList<>());
        return result;
    }


    private void helper(int[] candidates, int index, int target, List<Integer> path){
        //base
        if(target == 0){
            result.add(path);
            return;
        }
        if(target < 0 ||index == candidates.length) return;
        //logic
        // choose
        path.add(candidates[index]);
        helper(candidates, index, target - candidates[index], new ArrayList<>(path));
        path.remove(path.size() - 1);
        //not choose
        helper(candidates, index + 1, target, new ArrayList<>(path));   
    }
}

// Backtracking

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    result = list()
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        self.result = list()

        if candidates is None or len(candidates) == 0:
            return self.result
        
        self.helper(candidates, 0, target, path = list())

        return self.result
    
    def helper(self, candidates: List[int], index: int, target: int, path: List[int]) -> None:
        #base
        if target == 0:
            self.result.append(path.copy())
            return
        
        if target < 0:
            return 

        #logic
        for i in range(index, len(candidates)):
            path.append(candidates[i])
            self.helper(candidates, i, target - candidates[i], path)
            path.pop(-1)

#Backtrack For Loop recursion
# Time Complexity:
# O(n * 2^(m+n)) where m is the candidates length and n is the target

# Space Complexity:
# O(n * h) -> O(n^2)
